File: omni/providers/openai/appserver.py
"""A JSON-RPC client for ``codex app-server``.

One JSON object per line in both directions — no framing headers. Responses are
matched to requests by id; everything else is a notification and goes straight
to the listener. The server occasionally asks *us* something; anything omni does
not handle is declined politely so the server never waits on a dead end.
"""

import itertools
import json
import threading

from ...shared.proc import LineProcess

CLIENT = {"name": "silicon-omni", "version": "0.1.0"}


class AppServerError(RuntimeError):
    pass


class AppServer:
    def __init__(self, env=None, cwd=None, on_notify=None, argv=(), on_exit=None):
        self.argv = ["codex", "app-server", "--stdio", *argv]
        self.env = env
        self.cwd = cwd
        self.on_notify = on_notify
        self.on_exit = on_exit
        self.ids = itertools.count(1)
        self.pending: dict[int, list] = {}
        self.guard = threading.Lock()
        self.proc: LineProcess | None = None

    # ------------------------------------------------------------- transport

    def start(self, capabilities: dict | None = None) -> dict:
        self.proc = LineProcess(
            self.argv, on_line=self.line, env=self.env, cwd=self.cwd, on_exit=self.closed
        ).start()
        return self.call(
            "initialize", {"clientInfo": CLIENT, "capabilities": capabilities or {}}, timeout=30
        )

    def closed(self, code: int) -> None:
        """The server is gone: fail everything waiting on it rather than hang."""
        with self.guard:
            waiting, self.pending = self.pending, {}
        for slot in waiting.values():
            slot[0] = {"error": {"message": f"codex app-server exited with {code}"}}
            slot[1].set()
        if self.on_exit:
            self.on_exit(code)

    def line(self, text: str) -> None:
        try:
            message = json.loads(text)
        except json.JSONDecodeError:
            return
        if "method" in message and "id" in message:
            self.decline(message)
        elif "method" in message:
            if self.on_notify:
                self.on_notify(message["method"], message.get("params") or {})
        elif "id" in message:
            self.settle(message)

    def decline(self, message: dict) -> None:
        """A server-to-client request omni has no answer for."""
        self.write(
            {
                "jsonrpc": "2.0",
                "id": message["id"],
                "error": {"code": -32601, "message": f"{message['method']} not handled by omni"},
            }
        )

    def settle(self, message: dict) -> None:
        with self.guard:
            slot = self.pending.pop(message["id"], None)
        if slot:
            slot[0] = message
            slot[1].set()

    def write(self, message: dict) -> bool:
        return bool(self.proc and self.proc.send_line(json.dumps(message)))

    # --------------------------------------------------------------- calling

    def call(self, method: str, params=None, timeout: float = 120.0):
        request_id = next(self.ids)
        slot = [None, threading.Event()]
        with self.guard:
            self.pending[request_id] = slot
        body = {"jsonrpc": "2.0", "id": request_id, "method": method}
        if params is not None:
            body["params"] = params
        if not self.write(body):
            raise AppServerError(f"codex app-server is not running (wanted {method})")
        if not slot[1].wait(timeout):
            with self.guard:
                self.pending.pop(request_id, None)
            raise AppServerError(f"{method} timed out after {timeout}s")
        message = slot[0]
        if "error" in message:
            raise AppServerError(f"{method}: {message['error'].get('message', message['error'])}")
        return message.get("result")

    def try_call(self, method: str, params=None, timeout: float = 15.0):
        """For calls that are allowed to fail — interrupts, best-effort probes."""
        try:
            return self.call(method, params, timeout)
        except AppServerError:
            return None

    def notify(self, method: str, params=None) -> None:
        body = {"jsonrpc": "2.0", "method": method}
        if params is not None:
            body["params"] = params
        self.write(body)

    def stop(self) -> None:
        if self.proc:
            self.proc.stop(timeout=8)

    @property
    def alive(self) -> bool:
        return bool(self.proc and self.proc.alive)

File: crates/omni-core/src/providers/openai/appserver.rs
//! A JSON-RPC client for `codex app-server`.
//!
//! One JSON object per line in both directions — no framing headers. Responses
//! are matched to requests by id; everything else is a notification and goes
//! straight to the listener. The server occasionally asks *us* something;
//! anything omni does not handle is declined politely so the server never waits
//! on a dead end.

use std::collections::BTreeMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::{Arc, Mutex, mpsc};
use std::time::Duration;

use serde_json::{Value, json};

use crate::shared::proc::{LineProcess, Spawn};

const CLIENT: &str = "silicon-omni";

pub type OnNotify = Arc<dyn Fn(&str, &Value) + Send + Sync>;

#[derive(Debug)]
pub struct AppServerError(pub String);

impl std::fmt::Display for AppServerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

type Pending = Arc<Mutex<BTreeMap<i64, mpsc::Sender<Value>>>>;

pub struct AppServer {
    proc: Arc<LineProcess>,
    ids: AtomicI64,
    pending: Pending,
}

impl AppServer {
    /// Start the server and shake hands. `on_notify` hears everything after.
    pub fn start(
        argv: Vec<String>,
        env: Vec<(String, String)>,
        cwd: Option<String>,
        on_notify: OnNotify,
        on_exit: impl Fn(i32) + Send + Sync + 'static,
    ) -> Result<Self, AppServerError> {
        let pending: Pending = Arc::default();
        let heard = pending.clone();
        // Declining happens on the reader thread, which must not block on the
        // writer; it queues the refusal here and a writer thread sends it.
        let (decline_tx, decline_rx) = mpsc::channel::<Value>();
        let mut spawn = Spawn::new(argv).on_line(move |line| {
            let Ok(message) = serde_json::from_str::<Value>(line) else {
                return;
            };
            let method = message.get("method").and_then(Value::as_str);
            match (method, message.get("id")) {
                // A server-to-client request omni has no answer for.
                (Some(method), Some(id)) => {
                    let _ = decline_tx.send(json!({
                        "jsonrpc": "2.0",
                        "id": id,
                        "error": {"code": -32601, "message": format!("{method} not handled by omni")},
                    }));
                }
                (Some(method), None) => {
                    on_notify(method, message.get("params").unwrap_or(&Value::Null))
                }
                (None, Some(id)) => {
                    if let Some(id) = id.as_i64() {
                        let slot = heard.lock().unwrap_or_else(|p| p.into_inner()).remove(&id);
                        if let Some(tx) = slot {
                            let _ = tx.send(message);
                        }
                    }
                }
                _ => {}
            }
        });
        for (key, value) in env {
            spawn = spawn.env(&key, value);
        }
        if let Some(cwd) = cwd {
            spawn = spawn.cwd(cwd);
        }
        // The server is gone: fail everything waiting on it rather than hang.
        let orphaned = pending.clone();
        let proc = Arc::new(
            spawn
                .on_exit(move |code| {
                    let waiting =
                        std::mem::take(&mut *orphaned.lock().unwrap_or_else(|p| p.into_inner()));
                    for (_, tx) in waiting {
                        let _ = tx.send(json!({
                            "error": {"message": format!("codex app-server exited with {code}")}
                        }));
                    }
                    on_exit(code);
                })
                .start()
                .map_err(|err| {
                    AppServerError(format!("could not start codex app-server: {err}"))
                })?,
        );

        let writer = proc.clone();
        std::thread::spawn(move || {
            for reply in decline_rx {
                if !writer.send_line(&reply.to_string()) {
                    return;
                }
            }
        });

        let server = AppServer {
            proc,
            ids: AtomicI64::new(1),
            pending,
        };
        server.call(
            "initialize",
            json!({"clientInfo": {"name": CLIENT, "version": env!("CARGO_PKG_VERSION")}, "capabilities": {}}),
            Duration::from_secs(30),
        )?;
        Ok(server)
    }

    pub fn call(
        &self,
        method: &str,
        params: Value,
        timeout: Duration,
    ) -> Result<Value, AppServerError> {
        let id = self.ids.fetch_add(1, Ordering::SeqCst);
        let (tx, rx) = mpsc::channel();
        self.pending
            .lock()
            .unwrap_or_else(|p| p.into_inner())
            .insert(id, tx);
        let mut body = json!({"jsonrpc": "2.0", "id": id, "method": method});
        if !params.is_null() {
            body["params"] = params;
        }
        if !self.proc.send_line(&body.to_string()) {
            self.pending
                .lock()
                .unwrap_or_else(|p| p.into_inner())
                .remove(&id);
            return Err(AppServerError(format!(
                "codex app-server is not running (wanted {method})"
            )));
        }
        let Ok(message) = rx.recv_timeout(timeout) else {
            self.pending
                .lock()
                .unwrap_or_else(|p| p.into_inner())
                .remove(&id);
            return Err(AppServerError(format!(
                "{method} timed out after {timeout:?}"
            )));
        };
        if let Some(error) = message.get("error") {
            let said = error
                .get("message")
                .and_then(Value::as_str)
                .unwrap_or("failed");
            return Err(AppServerError(format!("{method}: {said}")));
        }
        Ok(message.get("result").cloned().unwrap_or(Value::Null))
    }

    /// For calls that are allowed to fail — interrupts, best-effort probes.
    pub fn try_call(&self, method: &str, params: Value, timeout: Duration) -> Option<Value> {
        self.call(method, params, timeout).ok()
    }

    pub fn stop(&self) {
        self.proc.stop(Duration::from_secs(8));
    }

    pub fn alive(&self) -> bool {
        self.proc.alive()
    }
}
